pages/apis/webhooks.md.erb

```diff
@@ -63,6 +64,39 @@ Each event’s data is sent JSON encoded in the request body. See each event’s
   <p>Note that if a builds transitions between states very quickly, for example from blocked (`finished`) to unblocked (`running`), the webhook may be in a different state from the actual build. This is a known limitation of webhooks, in that they may represent a later version of the object than the one that triggered the event.</p>
 </div>
 
+## Verifying Webhook signatures
+
+The `X-Buildkite-Signature` header included in each signed event contains a timestamp and one signature. The timestamp is prefixed by `timestamp=`, and each signature is prefixed by `signature=`.
+
+Buildkite generates signature using a hash-based message authentication code ([HMAC](https://en.wikipedia.org/wiki/HMAC)) with [SHA-256](https://en.wikipedia.org/wiki/SHA-2). The timestamp is an integer represents UTC timestamp.
+
+Example using Ruby:
+
+```ruby
+class BuildkiteWebhook
+  def self.valid?(payload, header, secret)
+    timestamp, signature = get_timestamp_and_signatures(header)
+    expected = OpenSSL::HMAC.hexdigest("sha256", secret, "#{timestamp}.#{payload}")
+    Rack::Utils.secure_compare(expected, signature)
+  end
+
+  def self.get_timestamp_and_signatures(header)
+    parts = header.split(",").map { |kv| kv.split("=", 2).map(&:strip) }.to_h
+    [parts["timestamp"], parts["signature"]]
+  end
+end
+
+BuildkiteWebhook.valid?(
+  request.body.read,
+  request.headers["X-Buildkite-Signature"],
+  ENV["BUILDKITE_WEBHOOK_SECRET"]
+)
+```
+
+### Preventing replay attacks
+
+A [replay attack](https://en.wikipedia.org/wiki/Replay_attack) is when an attacker intercepts a valid payload and its signature, then re-transmits them. To mitigate such attacks, Buildkite includes a timestamp in the `X-Buildkite-Signature` header. The timestamp is part of the signed payload, it is also verified by the signature, so an attacker cannot change the timestamp without invalidating the signature.
```

Comment:
To mitigate against a reply attack, does the webhook receiver need to:

1. verify the signature
2. check that the timestamp (which we can trust after #1) is within a reasonable timeframe (30s? 60s? 5 mins?)

Reply:
should i specify 5 minutes (or other timeframe) here?

Reply:
The approriate time frame might depend on the customers risk tolerance, but 5 minutes does seem like a pretty good suggestion to me